### PR TITLE
Add Magma Agentic Model from Microsoft

### PR DESCRIPTION
**Pull request by jwyang**
# What does this PR do?

Add [Microsoft Magma model](https://github.com/microsoft/Magma) to the repository

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
Yay, super excited to get the model shipped! I know it is early to review, I noticed the model doesn't have a modular file yet. I recommend to use [modular transformers](https://huggingface.co/docs/transformers/en/modular_transformers) to add the model, it will allow you to inherit from any similar model in transformers and you won't have to rewrite the whole class

Also it makes the review process easier and faster, since we see what are the main differences between Magma and other exusting model 😉

**Comment by jwyang** [comment]
@zucchini-nlp , thank for the pointer! First time hearing about the modular transformers. I will take a look.

File: src/transformers/models/magma/modeling_magma.py [hunk 1]
```diff
@@ -0,0 +1,908 @@
...
+
+
+class MagmaMultiModalProjector(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        
+        projector_type = config.mm_projector_type
+        mlp_gelu_match = re.match(r'^mlp(\d+)x_gelu$', projector_type)
+        if mlp_gelu_match:
```
Comment (zucchini-nlp) [inline src/transformers/models/magma/modeling_magma.py:97]: could there be no match? Maybe we can raise an error if that's not expected?

File: src/transformers/models/magma/modeling_magma.py [hunk 2]
```diff
@@ -0,0 +1,908 @@
...
+            mlp_depth = int(mlp_gelu_match.group(1))
+            modules = [nn.Linear(config.mm_hidden_size, config.hidden_size)]
+            for _ in range(1, mlp_depth):
+                modules.append(nn.GELU())
+                modules.append(nn.Linear(config.hidden_size, config.hidden_size))
+            self.proj = nn.Sequential(*modules)
```
Comment (zucchini-nlp) [inline src/transformers/models/magma/modeling_magma.py:103]: nit: we try to use `nn.ModuleList` in transformers instead of `Sequential`

File: src/transformers/models/magma/modeling_magma.py [hunk 3]
```diff
@@ -0,0 +1,908 @@
...
+
+        if "mm_use_row_seperator" not in config or config.mm_use_row_seperator:
+            # define a row seperator
+            self.row_seperator = nn.Parameter(torch.zeros(1, 1, config.hidden_size))
+        if 'mm_use_im_start_end' in config and config.mm_use_im_start_end:
+            self.img_start_seperator = nn.Parameter(torch.zeros(1, config.hidden_size))
+            self.img_end_seperator = nn.Parameter(torch.zeros(1, config.hidden_size))                        
+
```
Comment (zucchini-nlp) [inline src/transformers/models/magma/modeling_magma.py:111]: not used?

File: src/transformers/models/magma/modeling_magma.py [hunk 4]
```diff
@@ -0,0 +1,908 @@
...
+            self.deepspeed_engine.module = self
+        
+    def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_multiple_of=None) -> nn.Embedding:
+        model_embeds = self.language_model.resize_token_embeddings(new_num_tokens, pad_to_multiple_of)
+        # update vocab size
+        self.config.text_config.vocab_size = model_embeds.num_embeddings
+        self.vocab_size = model_embeds.num_embeddings
+        return model_embeds
+
+    def _merge_input_ids_with_image_features(
```
Comment (zucchini-nlp) [inline src/transformers/models/magma/modeling_magma.py:441]: should be removed. We now merge embeddings with `masked_scatter` by inserting placeholder tokens in processor

**Review by zucchini-nlp — COMMENTED** [review]
Hey @jwyang! I started reviewing as requested internally but seems like the PR is not yet ready for initial review

Model and image processing code looks very similar to existing llava-next, so I strongly recommend to import identical part with modular. And probably you'll need to rebase `main`, the current fork seems to be a bit outdated. Feel free to ping me under PR or internally, when it can be reviewed 🤗

File: src/transformers/models/magma/modeling_magma.py (hunk 1 above)
Comment (jwyang) [reply src/transformers/models/magma/modeling_magma.py:97]: good point!

File: src/transformers/__init__.py [hunk 5]
```diff
@@ -589,6 +589,11 @@
         "LxmertTokenizer",
     ],
     "models.m2m_100": ["M2M100Config"],
+    "models.magma": [
+        "MagmaConfig",
+        "MagmaProcessor",
+        "MagmaImageProcessor",
+    ],
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/__init__.py]: done

File: src/transformers/models/magma/configuration_magma.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    >>> # Initializing a model from the Magma-8B style configuration
+    >>> model = MagmaModel(configuration)
+
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
+
+    model_type = "magma"
+    keys_to_ignore_at_inference = ["past_key_values"]
+
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/configuration_magma.py:64]: done

File: src/transformers/models/magma/configuration_magma.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+
+    Args:
+        vision_config (`Union[AutoConfig, dict]`,  *optional*, defaults to `SiglipVisionConfig`):
+            The config object or dictionary of the vision backbone.
+        text_config (`Union[AutoConfig, dict]`, *optional*, defaults to `Qwen2Config`):
+            The config object or dictionary of the text backbone.
+        image_token_index (`int`, *optional*, defaults to 151646):
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/configuration_magma.py]: add the attribute mapping as in llava-onevision

File: src/transformers/models/magma/configuration_magma.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        vision_config=None,
+        text_config=None,
+        image_token_index=None,
+        tie_word_embeddings=False,
+        **kwargs,
+    ):
+        
+        if isinstance(vision_config, dict):
+            vision_config = PretrainedConfig(**vision_config)
+        self.vision_config = vision_config
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/configuration_magma.py:76]: done

File: src/transformers/models/magma/configuration_magma.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        
+        self.image_token_index = image_token_index
+
+        if isinstance(text_config, dict):
+            text_config["model_type"] = text_config["model_type"] if "model_type" in text_config else "llama"
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/configuration_magma.py:81]: good catch. It is llama text config

File: src/transformers/models/magma/configuration_magma.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            text_config = CONFIG_MAPPING[text_config["model_type"]](**text_config)
+        
+        if text_config is not None:
+            # copy all variables in text_config to self
+            for key, value in text_config.__dict__.items():
+                if not key.startswith("_") and not key.startswith("__"):
+                    setattr(self, key, value)
+            self.text_config = text_config
+        else:
+            self.text_config = None
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/configuration_magma.py:91]: good point, I recall that I used this to adapt to pure llm or vlm loading, will fix it

File: src/transformers/models/magma/image_processing_magma.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+            )
+                
+        if do_convert_rgb:
+            images = [convert_to_rgb(image) for image in images]
+  
+        # tensor transform and normalize
+        img_processor = torchvision.transforms.Compose([
+            torchvision.transforms.ToTensor(),
+            torchvision.transforms.Normalize(self.image_mean, self.image_std)
+        ])
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/image_processing_magma.py:234]: Should we use the slow or fast version? it seems that llava-next uses the slow image processor by default?

[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/magma/image_processing_magma.py:234]: For new models we usually add only fast processors, and fast will be loaded by default in the future releases

File: src/transformers/models/magma/configuration_magma.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/magma/configuration_magma.py]: can't we change the config? `attribute_map` is a hacky version we use for BC when the model's configs can't be updated anymore

File: src/transformers/models/magma/modeling_magma.py (hunk 4 above)
Comment (jwyang) [reply src/transformers/models/magma/modeling_magma.py]: thanks! fixed it now

File: src/transformers/models/magma/modeling_magma.py (hunk 3 above)
Comment (jwyang) [reply src/transformers/models/magma/modeling_magma.py:111]: we used row_seperator but not start/end seperator. Will remove the latter one.

File: src/transformers/models/magma/configuration_magma.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/configuration_magma.py]: I see, I will change the config directly then. When I worked with vLLM recently, I did notice it should be image_token_id, :).

File: src/transformers/models/magma/configuration_magma.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+    Args:
+        vision_config (`Union[AutoConfig, dict]`,  *optional*, defaults to `SiglipVisionConfig`):
+            The config object or dictionary of the vision backbone.
+        text_config (`Union[AutoConfig, dict]`, *optional*, defaults to `Qwen2Config`):
+            The config object or dictionary of the text backbone.
+        image_token_index (`int`, *optional*, defaults to 151646):
+            The image token index to encode the image prompt.
+        tie_word_embeddings (`bool`, *optional*, defaults to `False`):
+            Whether the model's input and output word embeddings should be tied.
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/configuration_magma.py:47]: I do have a tie_word_embeddings in my text config, should I make any change to the code here? As I follow other models, e.g., llava-onevision.

[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/magma/configuration_magma.py:114]: Hm I thought we cleaned up all models. Keeping it in general config has to effect so we can remove for Magma, and I will clean up other llava models

File: src/transformers/models/magma/image_tower_magma.py [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+        output_dict=output_dict,
+        **model_kwargs,
+    )
+
+class D2CLIP_HF(nn.Module):
+    def __init__(self, config, **kwargs):    
+        super().__init__()
+        if isinstance(config, PretrainedConfig):
+            self.model_name = config.vision_backbone
+        else:
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/image_tower_magma.py:277]: Hi, @zucchini-nlp, I am using openclip convnext backbone from here: hf-hub:laion/CLIP-convnext_xxlarge-laion2B-s34B-b82K-augreg. Seems that it is not supported yet with AutoModel. In this case, I may have to implement by myself by integrating openclip modules. Any suggestions?

[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/magma/image_tower_magma.py:277]: Looks like it should be loadable with timm backbone, cc @qubvel to confirm

[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/models/magma/image_tower_magma.py:277]: Yes, should be doable with one of the options 
1) ConvNext model in transformers
2) load_backbone ik case you need intermediate 2d feature maps 
3) AutoModel which should load TimmWrapper model

[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/image_tower_magma.py:277]: Got it, let me find a way which is most convinient among them. Thank you @zucchini-nlp @qubvel! Will let you know if I get any problems.

[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/image_tower_magma.py]: @zucchini-nlp @qubvel I am now using timm convnext to load the model, please have a check! 

https://github.com/jwyang/transformers/blob/714746dcd2cb5b584f17ad072f113a30655e81d0/src/transformers/models/magma/modeling_magma.py#L291

File: src/transformers/models/magma/image_processing_magma.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/image_processing_magma.py:204]: @zucchini-nlp , I added the fask image processing follow llava, please take a look!

https://github.com/jwyang/transformers/blob/main/src/transformers/models/magma/image_processing_magma_fast.py

File: src/transformers/models/magma/image_tower_magma.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/magma/image_tower_magma.py]: Cool, looks like we can use directly `AutoModel.from_pretrained` as pointed out above (documentation [here](https://huggingface.co/docs/transformers/en/model_doc/timm_wrapper#transformers.TimmWrapperModel))

[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/image_tower_magma.py]: @zucchini-nlp, using automodel to load convnext now. please take a look. It also passed all test.

[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/magma/image_tower_magma.py]: Perfect thanks! Looks good to me, also we'll need to infer model name from config instead of hardcoding a possible values

File: src/transformers/models/magma/modeling_magma.py [hunk 14]
```diff
@@ -0,0 +1,584 @@
...
+                    image_features_split = torch.split(image_features, num_crops_list, dim=0)
+                    flattened_image_features = []
+                    for image_feature, image_size in zip(image_features_split, _image_sizes_list_temp):
+                        image_feature = image_feature.view(image_size[0], image_size[1], *image_feature.shape[1:])
+                        image_feature = image_feature.permute(0, 2, 1, 3, 4).flatten(2, 3).flatten(0, 1)
+                        if self.config.vision_config.mm_use_row_seperator:
+                            image_feature = torch.cat((image_feature, self.multi_modal_projector.row_seperator.repeat(image_feature.shape[0],1,1)), dim=1)
+                        flattened_image_features.append(image_feature.flatten(0, 1))
+
+                inputs_embeds[input_ids == self.config.image_token_id] = torch.cat(flattened_image_features, dim=0)
```
Comment (zucchini-nlp) [inline src/transformers/models/magma/modeling_magma.py:472]: logic to obtain `flattened_image_features` can be abstracted away to `self.get_image_features`. That will help us a lot in integrating with text inference frameworks and keeping a standard format in VLMs

**Review by zucchini-nlp — COMMENTED** [review]
@jwyang hey, thanks a lot for iterating and using the modular!

I see the modular conversion wasn't done yet, and converting with the current modular file would wipe out the vision part and magma-specific code. You can convert the model with `python utils/modular_model_converter.py` and check the auto-generated code. If some parts have to be different, then only the modular file is modified

Overall, I left comments on the general format of code. We can push a bit more to standardize and use the best practices for VLMs and not rely on `remote_code` when loading. That will help us in integrating with TGI/vLLM/sglang etc easier, if all models follow same format.

The format I expect is for model to have a base `MagmaModel` without an LM head and `MagmaForCausalLM` which adds a head on top. The base model will also need to have a few helper methods like `get_image_features`

I am not sure if you are open to updating model weights/configs on the hub though. Because the above comments will require a conversion script. LMK if the hub repo can't be modified for any reason, I will see what we can do 🤗

File: src/transformers/models/magma/image_processing_magma.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/image_processing_magma.py:204]: Hi, @zucchini-nlp , when I deleted the slow one, the following code:

`processor = AutoProcessor.from_pretrained("microsoft/Magma-8B")`

will cause error, and can be solved only by adding "use_fast=True". 

Though my transformers say "Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor." but it seems that the AutoProcessor does not automatically load the fast version by default.

Any idea?

File: src/transformers/models/magma/image_tower_magma.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/image_tower_magma.py]: I presume this is resolved

File: src/transformers/models/__init__.py [hunk 15]
```diff
@@ -13,6 +13,317 @@
 # limitations under the License.
 from typing import TYPE_CHECKING
 
+from . import (
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/__init__.py]: thank for reminding this! It was indeed a faulty merge. Fixed!

File: src/transformers/models/magma/modular_magma.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+@dataclass
+class MagmaCausalLMOutputWithPast(LlavaNextCausalLMOutputWithPast):
+    """
+    video_hidden_states (`torch.FloatTensor`, *optional*):
+        A `torch.FloatTensor`  of size `(batch_size * num_frames, num_videos, sequence_length, hidden_size)`.
+        video_hidden_states of the model produced by the vision encoder and after projecting the last hidden state.
+    """
+    pass
+
+class MagmaPreTrainedModel(LlavaNextPreTrainedModel):
```
[the remark this answers is not in the document]
Comment (jwyang) [reply src/transformers/models/magma/modular_magma.py:255]: I see, now it becomes more and more clearer to me on how this modular file is used, :)
